src/uct/tcp/tcp_cm.c

```diff
@@ -8,86 +8,329 @@
 #include <ucs/async/async.h>
 
 
-unsigned uct_tcp_cm_conn_progress(uct_tcp_ep_t *ep)
+void uct_tcp_cm_change_conn_state(uct_tcp_ep_t *ep,
+                                  uct_tcp_ep_conn_state_t new_conn_state)
 {
     uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
                                             uct_tcp_iface_t);
+    char str_local_addr[UCS_SOCKADDR_STRING_LEN];
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
+    uct_tcp_ep_conn_state_t old_conn_state;
+
+    old_conn_state = ep->conn_state;
+    ep->conn_state = new_conn_state;
+
+    switch(ep->conn_state) {
+    case UCT_TCP_EP_CONN_STATE_CONNECTING:
+    case UCT_TCP_EP_CONN_STATE_WAITING_ACK:
+        if (old_conn_state == UCT_TCP_EP_CONN_STATE_CLOSED) {
+            uct_tcp_iface_outstanding_inc(iface);
+        } else {
+            ucs_assert(old_conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING);
+        }
+        break;
+    case UCT_TCP_EP_CONN_STATE_CONNECTED:
+        ucs_assert((old_conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK) ||
+                   (old_conn_state == UCT_TCP_EP_CONN_STATE_ACCEPTING));
+        if (old_conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK) {
+            uct_tcp_iface_outstanding_dec(iface);
+        }
+        break;
+    case UCT_TCP_EP_CONN_STATE_CLOSED:
+        ucs_assert(old_conn_state != UCT_TCP_EP_CONN_STATE_CLOSED);
+        if ((old_conn_state == UCT_TCP_EP_CONN_STATE_CONNECTING) ||
+            (old_conn_state == UCT_TCP_EP_CONN_STATE_WAITING_ACK)) {
+            uct_tcp_iface_outstanding_dec(iface);
+        }
+        break;
+    default:
+        ucs_assert(ep->conn_state == UCT_TCP_EP_CONN_STATE_ACCEPTING);
+        break;
+    }
+
+    ucs_debug("tcp_ep %p: %s -> %s for the [%s]<->[%s] connection",
+              ep, uct_tcp_ep_cm_state[old_conn_state].name,
+              uct_tcp_ep_cm_state[ep->conn_state].name,
+              ucs_sockaddr_str((const struct sockaddr*)&iface->config.ifaddr,
+                               str_local_addr, UCS_SOCKADDR_STRING_LEN),
+              ucs_sockaddr_str(ep->peer_addr->addr, str_remote_addr,
+                               UCS_SOCKADDR_STRING_LEN));
+}
+
+static void uct_tcp_cm_trace_conn_pkt(const uct_tcp_ep_t *ep, const char *msg,
+                                      const struct sockaddr *peer_addr)
+{
+    char str_addr[UCS_SOCKADDR_STRING_LEN];
+
+    ucs_debug("tcp_ep %p: %s %s", ep, msg,
+              ucs_sockaddr_str(peer_addr, str_addr,
+                               UCS_SOCKADDR_STRING_LEN));
+}
+
+static ucs_status_t
+uct_tcp_cm_conn_pkt_check_event(const uct_tcp_ep_t *ep,
+                                uct_tcp_cm_conn_event_t expected_event,
+                                uct_tcp_cm_conn_event_t actual_event)
+{
+    char str_addr[UCS_SOCKADDR_STRING_LEN];
+    ucs_assertv_always(expected_event == actual_event,
```

Comment:
why is this function always returning UCS_OK?
it should either be void, or return error instead of asserting
i think it should be ucs_error since assert is a "local" error, and we in general should not crash the program in case of protocol violation